crates/router/src/core/payments/operations/payment_confirm.rs

```diff
@@ -1093,6 +1094,110 @@ impl<F: Clone + Send + Sync> Domain<F, api::PaymentsRequest, PaymentData<F>> for
         Ok(())
     }
 
+    async fn apply_three_ds_authentication_strategy<'a>(
+        &'a self,
+        state: &SessionState,
+        payment_data: &mut PaymentData<F>,
+        business_profile: &domain::Profile,
+    ) -> CustomResult<(), errors::ApiErrorResponse> {
+        // If the business profile has a three_ds_decision_rule_algorithm, we will use it to determine the 3DS strategy (authentication_type, exemption_type and force_three_ds_challenge)
+        if let Some(three_ds_decision_rule) =
+            business_profile.three_ds_decision_rule_algorithm.clone()
+        {
+            // Parse the three_ds_decision_rule to get the algorithm_id
+            let algorithm_id = three_ds_decision_rule
+                .parse_value::<api::routing::RoutingAlgorithmRef>("RoutingAlgorithmRef")
+                .change_context(errors::ApiErrorResponse::InternalServerError)
+                .attach_printable("Could not decode merchant routing algorithm ref")?
```

Comment:
Nit: Should it be profile instead of merchant?

```suggestion
                .attach_printable("Could not decode profile routing algorithm ref")?
```